### DISCUSSION TITLE
Ciphertrust Azure Implementation Question

### DISCUSSION THREAD

**Discussion by dawsonar802** [ThalesGroup/terraform-provider-ciphertrust#3] [General]
Hello, 
I am looking at using this provider to implement a connection to CM and also use an exiting key and key vault.  While reviewing the documentation, it was not clear to me on the following items.

1. When creating the connection, we are already leveraging the ARM_ environment variables for the TF SPN, but the Thales CCKM has a separate SPN that would need to be used.  Would it be best to leverage the "Create an Azure connection without using environment variables" option from the docs?  Or if using the existing ARM_ environment variables, what rights are needed for the SPN to make the connection?
2. Is it possible to leverage existing Vaults and Keys?  Meaning we are provisioning a Key vault already via TF and would like to leave that as is.  We would want to use an existing key that is already created in the CIphertrust manager (on-prem).  The goal being to deploy the normal azure key vault via Terraform and then add/import a key into the vault that has already been created.  Is this possible or does the key need to be created using the `ciphertrust_azure_key` resource?  
3. My understanding is that the `ciphertrust_azure_vault` is not creating the vault, it is only creating a ciphertrust resource for an existing azurerm vault, is that correct?  The same with `ciphertrust_azure_key` it appears I could use this to either create a native key or upload an existing local/on-prem key.  If uploading, where can the `local_key_id ` be found, could that be found using the `ciphertrust_azure_connection` resource?

I apologize for all of the questions, but it looks like this is a fairly new provider and I just want to be sure I am understanding everything prior to trying to implement.  

Thanks.
dawsonar802

**Comment by SarahThompson** [comment]
Hi,

Thank-you for the inquiry. 

With respect to your questions:

Question 1. Please refer to https://thalesdocs.com/ctp/cm/latest/admin/cckm_ag/azure/azure-prereqs/azure-cloud-prereq/index.html on how to register and configure an application for use with CipherTrust.

If the credentials have to differ from those specified in the ARM environment variables then, yes, specify them in the terraform script. 
resource "ciphertrust_azure_connection" "azure_connection" {
  name          = "azure-connection-name"
  client_id     = "client-id"
  client_secret = "client-secret"
  tenant_id     = "tenant-id"
}

Where the above values are:
client_id:   "Application (client) ID" found in the application's properties
tenant_id: "Directory (tenant) ID" found in the application's properties.
client_secret: A secret created in the Certificates & secrets page of the application.

Questions 2. and 3. 
I think to start with referring to the following pages might help in understanding the flow.
https://github.com/ThalesGroup/terraform-provider-ciphertrust/tree/main/sample-scripts/azure-keys
https://github.com/ThalesGroup/terraform-provider-ciphertrust/tree/main/sample-scripts/azure-keys/upload/ciphertrust
https://github.com/ThalesGroup/terraform-provider-ciphertrust/blob/main/sample-scripts/azure-keys/upload/ciphertrust/main.tf

If you have further inquiries please don't hesitate to ask.

Regards Sarah

**Comment by dawsonar802** [comment]
Thank you for the reply.  With regards to question number 1, are there specific permissions within CM or in Azure (RBAC) that the CM SPN would need, I am not seeing this documented anywhere.  For example, with out current SPN we would allow it Owner or Contribute to be able to deploy the resources into Azure, but it doesn't have access to the local CM on-prem.  

With regards to questions 2 and 3, am I correct in thinking the the `ciphertrust_cm_key` resource would create a key in the CM (on-prem)?  From there I see there is an option to upload it to a vault using `ciphertrust_azure_key`, but again I am not sure if the vault is one that can be specified or if it needs to be created using the `ciphertrust_azure_vault`.  

Thanks again.

**Reply by SarahThompson** [reply]
With regards to Question 2. Yes, creating a ciphertrust_cm_key resource will create a key in your CM, and yes, the vault needs to be "created" using ciphertrust_azure_vault and the resource id specified in the ciphertrust_azure_key to upload the key to the vault. ciphertrust_azure_vault  doesn't create a vault in azure it just creates a vault resource in CM for the vault specified in the resource.  

With regards to Question 1. Sorry, I'm not really following. I think to get those questions answered you need pre-sales support. I think even if you were not using terraform you would still have to same questions? Hopefully this page will lead you do the right place for that support. https://cpl.thalesgroup.com/contact-us.

Regards Sarah

**Reply by SarahThompson** [reply]
Perhaps this area of the documentation might be of some help: https://thalesdocs.com/ctp/cm/2.8/admin/cckm_ag/azure/azure-prereqs/azure-stack-adfs-prereq/index.html

**Comment by dawsonar802** [comment]
Excellent, thank you that helps.  Yes, I agree I may still have some questions, but this should get me started.  I found the following from the link you posted as well which should help with the prerequisites needed for the SPN: https://thalesdocs.com/ctp/cm/2.8/admin/cckm_ag/azure/azure-prereqs/azure-cloud-prereq/index.html#set-required-permissions-on-cckm-app-on-azure-portal

Thanks again, I look forward to giving this provider a try.

**Comment by dawsonar802** [comment]
Sorry, one additional question on the provider.  It looks like in addition to the SPN information (Client ID and Secret), I need to supply the `CM_USERNAME` and `CM_PASSWORD`?  Since I do not have access to the on-prem CM, is this something that the client can just create a new user for with say read access or is there specific access needed for this user?  When I try with the client id/secret I get an error: `Error: The terraform-provider-ciphertrust_v0.9.0-beta8 plugin crashed!`

**Comment by SarahThompson** [comment]
Hi, 
Terribly sorry for the delayed reply and sorry the provider crashed :(  
If you can add log_level="debug" to your provider block it might help in debugging the crash should it happen again.  The default log file is ctp.log.
When you say "I do not have access to the on-prem CM" you have network access right? 
There is always a CM admin user do you know the password for admin?
Initially the admin user can create another user and assign them to which ever groups are required for instance CCKM Admins and affiliated groups or even the admin group.
The following pages might help navigate theses issues.
https://www.thalesdocs.com/ctp/cm/latest/admin/cm_admin/users/index.html
https://www.thalesdocs.com/ctp/cm/latest/admin/cm_admin/groups/index.html
Regards Sarah.

**Comment by dawsonar802** [comment]
No problem.  

> "When you say "I do not have access to the on-prem CM" you have network access right?"

Yes, network access should be fine.  We are assisting a customer with testing the Terraform module to leverage existing keys in CM, so I do not currently have access (via an account) to CM.  I am in the process of getting this, so will be able to try again soon.  

I will also set the log level to debug if I receive the error again.  Thanks.

**Comment by dawsonar802** [comment]
@SarahThompson , 
I tried setting the log_level to "debug", however it seems like no file is being created.  It is like it is failing before it can generate the file possibly.  I have tried various combinations of the provider beta5, beta7, etc..., which seem to give the same error.  I have updated the Terraform version to the latest as we were using 1.1.2, but still the same error.  

I am using a module with the `ciphertrust_azure_connection` resource in it, and this is running through Terraform Cloud, but other than that the configuration is extremely basic.  



## Module
```
terraform {
  required_providers {
    ciphertrust = {
      source  = "ThalesGroup/ciphertrust"
      version = "0.9.0-beta8"
    }
  }
}

provider "ciphertrust" {
  log_level = "debug"
  log_file  = "ctp.log"
}

data "azurerm_client_config" "current" {}

resource "ciphertrust_azure_connection" "azure_connection" {
  name          = var.thales.connection_name
  client_id     = var.thales.client_id
  client_secret = var.thales.client_secret
  tenant_id     = data.azurerm_client_config.current.tenant_id
}

variable "thales" {

}
```

## Call to module

```
module "thales" {
  source = "../../../../modules/thales"

  thales = {
    connection_name = "poc-testing-cm"
    client_id       = var.thales_client_id
    client_secret   = var.thales_client_secret

  }

}
```

Then within Terraform Cloud we have the client id and secret as terraform variables and the CM_USERNAME and CM_PASSWORD as environment variables.

**Comment by dawsonar802** [comment]
So I found a couple things that I now need to look into a bit further.  While we are allowing access from Azure to on-prem and on-prem to Azure, it appears that there is some isolation on the CM that was configured so the Azure vnet/subnets are not able to access it.  In addition it appears DNS resolution will be needed for the CM_ADDRESS host to be leveraged.  

Thank you again for you help on this and I will reach back out should I come up with any more questions on this.
